corehq/apps/data_analytics/malt_generator.py

```diff
@@ -79,19 +82,36 @@ def _save_to_db(cls, malt_rows_to_save, domain_id):
         except IntegrityError:
             # no update_or_create in django-1.6
             for malt_dict in malt_rows_to_save:
-                try:
-                    unique_field_dict = {k: v
-                                         for (k, v) in malt_dict.iteritems()
-                                         if k in MALTRow.get_unique_fields()}
-                    prev_obj = MALTRow.objects.get(**unique_field_dict)
-                    for k, v in malt_dict.iteritems():
-                        setattr(prev_obj, k, v)
-                    prev_obj.save()
-                except MALTRow.DoesNotExist:
-                    MALTRow(**malt_dict).save()
+                cls._update_or_create(malt_dict)
+        except Exception as ex:
+            logger.error("Failed to insert rows for domain with id {id}. Exception is {ex}".format(
+                         id=domain_id, ex=str(ex)), exc_info=True)
+
+    @classmethod
+    def _update_or_create(cls, malt_dict):
+        try:
+            # try update
+            unique_field_dict = {k: v
+                                 for (k, v) in malt_dict.iteritems()
+                                 if k in MALTRow.get_unique_fields()}
+            prev_obj = MALTRow.objects.get(**unique_field_dict)
+            for k, v in malt_dict.iteritems():
+                setattr(prev_obj, k, v)
+            prev_obj.save()
+        except MALTRow.DoesNotExist:
+            # create
+            try:
+                MALTRow(**malt_dict).save()
+            except Exception as ex:
+                logger.error("Failed to insert malt-row {}. Exception is {}".format(
+                    str(malt_dict),
+                    str(ex)
+                ), exc_info=True)
```

Comment:
Not a blocker, and I realize you're following the pattern used earlier in this file, but `str.format` is not as efficient as using `%s`-style formatting for logging, like so:
```py
logger.error("Failed to insert malt-row %s. Exception is %s", malt_dict, ex, exc_info=True)
```
Notice that no string formatting occurs in that statement (the logger does it internally if/when it needs to). If logging output is disabled then string formatting never occurs, and that reduces runtime overhead for disabled logging calls. In this case it really doesn't matter, but for cases like debug logging it can make quite a difference.